File: Graphs/graph-dfs.java
// DFS - Graph traversal
// Time: O(V + E)
// Space: O(V)

import java.util.*;

public class GraphDFS {
    private Map<Integer, List<Integer>> graph = new HashMap<>();
    private Set<Integer> visited = new HashSet<>();

    public void dfs(int node) {
        visited.add(node);
        System.out.print(node + " ");
        for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
            if (!visited.contains(neighbor)) {
                dfs(neighbor);
            }
        }
    }
}

File: Graphs/graph-dfs.c
// DFS - Graph traversal
// Time: O(V + E)
// Space: O(V)

#include <stdio.h>
#include <stdbool.h>

#define MAX_VERTICES 100

int graph[MAX_VERTICES][MAX_VERTICES]; // adjacency matrix
bool visited[MAX_VERTICES];
int V; // number of vertices

void dfs(int node) {
    visited[node] = true;
    printf("%d ", node);
    for (int i = 0; i < V; i++) {
        if (graph[node][i] && !visited[i]) {
            dfs(i);
        }
    }
}
